pkg/virtualgarden/kube_api_server_secrets.go

```diff
@@ -0,0 +1,380 @@
+	})
+	if err != nil {
+		return "", err
+	}
+
+	checksums[ChecksumKeyKubeAPIServerBasicAuth] = utils.ComputeChecksum(secret.Data)
+	return returnPw, nil
+}
+
+func (o *operation) deployKubeApiServerSecretEncryptionConfig(ctx context.Context, checksums map[string]string) error {
+	const encryptionConfigKey = "encryption-config.yaml"
+
+	var encryptionConfigValue []byte
+
+	secret := o.emptySecret(KubeApiServerSecretNameEncryptionConfig)
+	err := o.client.Get(ctx, client.ObjectKeyFromObject(secret), secret)
+	if err != nil {
+		if client.IgnoreNotFound(err) != nil {
+			return err
+		}
+
+		// secret does not exist: generate encryption config
+		encryptionConfigValue, err = o.generateNewEncryptionConfig()
+		if err != nil {
+			return err
+		}
+	} else {
+		// secret exists: use existing value
+		encryptionConfigValue = secret.Data[encryptionConfigKey]
+	}
+
+	_, err = controllerutil.CreateOrUpdate(ctx, o.client, secret, func() error {
+		if secret.Data == nil {
+			secret.Data = make(map[string][]byte)
+		}
+		secret.Data[encryptionConfigKey] = encryptionConfigValue
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	checksums[ChecksumKeyKubeAPIServerEncryptionConfig] = utils.ComputeChecksum(secret.Data)
+	return nil
+}
+
+func (o *operation) generateNewEncryptionConfig() ([]byte, error) {
+	secretBytes := make([]byte, 32)
+	if _, err := cryptorand.Read(secretBytes); err != nil {
+		return nil, err
+	}
+
+	secretString := utils.EncodeBase64(secretBytes)
+
+	encryptionConfig := configv1.EncryptionConfiguration{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "EncryptionConfiguration",
+			APIVersion: "apiserver.config.k8s.io/v1",
+		},
+		Resources: []configv1.ResourceConfiguration{
+			{
+				Resources: []string{
+					"secrets",
+				},
+				Providers: []configv1.ProviderConfiguration{
+					{
+						AESCBC: &configv1.AESConfiguration{
+							Keys: []configv1.Key{
+								{
+									Name:   "key",
+									Secret: secretString,
+								},
+							},
+						},
+					},
+					{
+						Identity: &configv1.IdentityConfiguration{},
+					},
+				},
+			},
+		},
+	}
+
+	return yaml.Marshal(&encryptionConfig)
+}
+
+func (o *operation) deployKubeApiServerSecretServiceAccountKey(ctx context.Context, checksums map[string]string) error {
+	const key = "service_account.key"
+
+	certConfig := &secretsutil.CertificateSecretConfig{
+		Name:       KubeApiServerSecretNameServiceAccountKey,
+		CertType:   secretsutil.CACert,
+		CommonName: Prefix + ":ca:kube-apiserver",
+		DNSNames: []string{
+			"virtual-garden:ca:kube-apiserver",
+		},
+	}
+
+	var value []byte
+
+	secret := o.emptySecret(KubeApiServerSecretNameServiceAccountKey)
+	err := o.client.Get(ctx, client.ObjectKeyFromObject(secret), secret)
+	if err != nil {
+		if client.IgnoreNotFound(err) != nil {
+			return err
+		}
+
+		// secret does not exist: generate certificate
+		cert, err := certConfig.GenerateCertificate()
+		if err != nil {
+			return err
+		}
+
+		value = cert.PrivateKeyPEM
```

Comment:
We tried to get rid the state directory in the git repo, and store the keys and certificates only in the cluster. 
What would happen if we lose the service account RSA private key and a new one is generated?

Reply:
I am not sure. Would have to investigate or try out if the old service account tokens still work.

Reply:
We have installed the virtual garden, then deleted the `virtual-garden-service-account-key`secret and restarted the pods of the virtual garden. The pods (virtual-garden-kube-apiserver and virtual-garden-kube-controller-manager) did not start.

Then we executed the deployment of the virtual garden again, so that a new `virtual-garden-service-account-key`secret  was generated. The pods started, and the kube apiserver could be accessed, for example to read namespaces.

I guess that we do not need to store the secret somewhere else, since it could be recreated.

Reply:
This private RSA key will be set in the flag `--service-account-key-file` right?
Usually, this contains the public key to verify the service account tokens generated by the KCM. In this case, it contains a private key and the API server extracts the public key from it for convenience.

If that private key is changed, I suppose the API server will fail to validate the tokens issued by the GCM.
As such this is very valuable :) 

In our current setup we supply the same RSA private key as used for the TLS serving certificate, but this is not necessary (even discouraged).

I suppose this should be exported in case someone wants to save this information.
In short: it depends whether we want to export these secrets to be stored externally. The secret itself is important.

Reply:
I talked to @schrodit about it, and we agreed that we would be most flexible if we provide both the means to optionally import and export this RSA private key.

> We tried to get rid the state directory in the git repo, and store the keys and certificates only in the cluster.

Opened for this one: https://github.com/gardener/landscaper/issues/307
On the agenda for an upcoming sync. 
But for now, I would implement it as described above if you are fine with it.

Reply:
I'm not sure if I understand your last paragraph correctly. Should we store the secret externally or not? As we could recreate it all the time I do not understand why?

Reply:
> As we could recreate it all the time I do not understand why

If you re-generate the service account private key the pods running in the cluster will break on the next KAPI restart. So no, you cannot.

> Should we store the secret externally or not

It is currently only stored in a secret resource (which is problematic  gardener/landscaper#307) .
I suggest for now to export it and import it.  A landscaper mechanism or a dedicated component then can take care of backing up that state and supplying it as import configuration in a subsequent run.

Reply:
> Then we executed the deployment of the virtual garden again, so that a new virtual-garden-service-account-keysecret was generated. The pods started, and the kube apiserver could be accessed, for example to read namespaces.

That does not seem to test the relevant part. Sure the GAPI will fail to start if the secret is not provided & when it starts again with a new service account token key file it will be acccessible again.

Though what should be tested is whether service accounts signed by the old RSA private key in the Gardener controller manager (& verified in the API server) can still be used against the Virtual Garden API server when the service account private key has changed.
I currently do not have my own virtual garden setup to try it out.

Reply:
This means:
- We will introduce a new optional parameter for the key. If the import parameter is provided, we use it; otherwise we generate a new key.
- The key is returned in a new export parameter. (This is either the imported or the new generated key.)
- Issue [Strategy for Backup Handling #307](https://github.com/gardener/landscaper/issues/307) must clarify where the  exported value is stored and the import value comes from.

Okay?

Reply:
I just validated that my assumption are correct on  another cluster

I updated the secret holding the service account key with another RSA private key.
Then I restart the API server.

Accessing the API with default token before rotation works

```
(shoot--dev--aws-clone:default) ~  curl -X GET https://api.aws-clone.dev.shoot.dev.k8s-hana.ondemand.com/api --header "Authorization: Bearer $TOKEN" --insecure
{
  "kind": "APIVersions",
  "versions": [
    "v1"
  ],
  "serverAddressByClientCIDRs": [
    {
      "clientCIDR": "0.0.0.0/0",
      "serverAddress": "api.aws-clone.dev.internal.dev.k8s.ondemand.com:443"
    }
  ]
}%
```

Unauthorized after the rotation

```
(shoot--dev--aws-clone:default) ~  curl -X GET https://api.aws-clone.dev.shoot.dev.k8s-hana.ondemand.com/api --header "Authorization: Bearer $TOKEN" --insecure
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {

  },
  "status": "Failure",
  "message": "Unauthorized",
  "reason": "Unauthorized",
  "code": 401
}%
```

All pods mounting service account tokens are failing with

> rent data during resync ListRoot="/calico/resources/v3/projectcalico.org/nodes" error=connection is unauthorized: Unauthorized
E0909 14:37:56.269643       1 reflector.go:123] pkg/mod/k8s.io/client-go@v0.0.0-20191114101535-6c5935290e33/tools/cache/reflector.go:96: Failed to list *v1.Node: Unauthorized
2021-09-09 14:37:56.714 [INFO][1] watchercache.go 206: Failed to perform list of current data during resync ListRoot="/calico/resources/v3/projectcalico.org/nodes" error=connection is unauthorized: Unauthorized
E0909 14:37:57.301366       1 reflector.go:123] pkg/mod/k8s.io/client-go@v0.0.0-20191114101535-6c5935290e33/tools/cache/reflector.go:96: Failed to list *v1.Node: Unauthorized
2021-09-09 14:37:57.746 [INFO][1] watchercache.go 206: Failed to perform list of current data during resync ListRoot="/calico/resources/v3/projectcalico.org/nodes" error=connection is unauthorized: Unauthorized
E0909 14:37:58.333082       1 reflector.go:123] pkg/mod/k8s.io/client-go@v0.0.0-20191114101535-6c5935290e33/tools/cache/reflector.go:96: Failed to list *v1.Node: Unauthorized
2021-09-09 14:37:58.778 [INFO][1] watchercache.go 206: Failed to perform list of current data during resync ListRoot="/calico/resources/v3/projectcalico.org/nodes" error=connection is unauthorized: Unauthorized
E0909 14:37:59.364796       1 reflector.go:123] pkg/mod/k8s.io/client-go@v0.0.0-20191114101535-6c5935290e33/tools/cache/reflector.go:96: Failed to list *v1.Node: Unauthorize


Also the API server logs are clear:

```
E0909 14:35:33.050931       1 authentication.go:63] "Unable to authenticate the request" err="[invalid bearer token, square/go-jose: error in cryptographic primitive]"
E0909 14:35:33.052825       1 authentication.go:63] "Unable to authenticate the request" err="[invalid bearer token, square/go-jose: error in cryptographic primitive]"
E0909 14:35:33.052828       1 authentication.go:63] "Unable to authenticate the request" err="[invalid bearer token, square/go-jose: error in cryptographic primitive]
```

Reply:
> We will introduce a new optional parameter for the key. If the import parameter is provided, we use it; otherwise we generate a new key.
The key is returned in a new export parameter. (This is either the imported or the new generated key.)
Issue Strategy for Backup Handling #307 must clarify where the exported value is stored and the import value comes from.
Okay?


Sound good, thanks.

Reply:
I have implemented the changes concerning the service account key as discussed. 
I added a validation for the imported key (if it is provided).

Reply:
ok great, will test it later today.